google-ads-examples/src/main/java/com/google/ads/googleads/examples/shoppingads/AddPerformanceMaxProductListingGroupTree.java

```diff
@@ -0,0 +1,556 @@
+        // correct order (child listing group filters must be removed before their parents).
+        AssetGroupListingGroupFilterRemoveOperationFactory removeOperationFactory =
+            new AssetGroupListingGroupFilterRemoveOperationFactory(existingListingGroupFilters);
+
+        operations.addAll(removeOperationFactory.removeAll());
+      }
+    }
+
+    // Uses a factory to create all the MutateOperations that manipulate a specific
+    // AssetGroup for a specific customer. The operations returned by the factory's methods
+    // are used to construct a new tree of filters. These filters can have parent-child
+    // relationships, and also include a special root that includes all children.
+    //
+    // When creating these filters, temporary IDs are used to create the hierarchy between
+    // each of the nodes in the tree, beginning with the root listing group filter.
+    //
+    // The factory created below is specific to a customerId and assetGroupId.
+    AssetGroupListingGroupFilterCreateOperationFactory createOperationFactory =
+        new AssetGroupListingGroupFilterCreateOperationFactory(
+            customerId, assetGroupId, TEMPORARY_ID_LISTING_GROUP_ROOT);
+
+    // Creates the operation to add the root node of the tree.
+    operations.add(createOperationFactory.createRoot());
+
+    // Creates an operation to add a leaf node for new products.
+    ListingGroupFilterDimension newProductDimension =
+        ListingGroupFilterDimension.newBuilder()
+            .setProductCondition(
+                ProductCondition.newBuilder()
+                    .setCondition(ListingGroupFilterProductCondition.NEW)
+                    .build())
+            .build();
+    operations.add(
+        createOperationFactory.createUnit(
+            TEMPORARY_ID_LISTING_GROUP_ROOT,
+            AssetGroupListingGroupFilterCreateOperationFactory.idGenerator.next(),
+            newProductDimension));
+
+    // Creates an operation to add a leaf node for used products.
+    ListingGroupFilterDimension usedProductDimension =
+        ListingGroupFilterDimension.newBuilder()
+            .setProductCondition(
+                ProductCondition.newBuilder()
+                    .setCondition(ListingGroupFilterProductCondition.USED)
+                    .build())
+            .build();
+    operations.add(
+        createOperationFactory.createUnit(
+            TEMPORARY_ID_LISTING_GROUP_ROOT,
+            AssetGroupListingGroupFilterCreateOperationFactory.idGenerator.next(),
+            usedProductDimension));
+
+    // This represents the ID of the "other" category in the ProductCondition subdivision. This ID
+    // is saved because the node with this ID will be further partitioned, and this ID will serve as
+    // the parent ID for subsequent child nodes of the "other" category.
+    long otherSubdivisionId = AssetGroupListingGroupFilterCreateOperationFactory.idGenerator.next();
```

Comment:
this feels a little weird to get the private idGenerator property of the factory object because it breaks encapsulation. Consider making a public method on the factory that calls the next method, or exposes the idGenerator.

Reply:
Agreed and fixed.